Gillian-C/lib/gilgen.ml

```diff
@@ -130,18 +130,110 @@ let trans_binop_expr ~fname binop te1 te2 =
+  match expr with
+  | Evar id -> Fmt.pf fmt "Evar id NUM[%d]" (num_to_integer id)
+  | Eaddrof _id -> Fmt.pf fmt "Eaddrof id"
+  | Econst _constant -> Fmt.pf fmt "Econst constant"
+  | Eunop (_unop, e) -> Fmt.pf fmt "Eunop unop (%a)" pp_expr e
+  | Ebinop (_binop, e1, e2) ->
+      Fmt.pf fmt "Ebinop binop (%a) (%a)" pp_expr e1 pp_expr e2
+  | Eload (_memory_chunk, e) -> Fmt.pf fmt "Eload mem_chunk (%a)" pp_expr e
+
+let rec pp_stmt fmt stmt =
+  match stmt with
+  | Sskip -> Fmt.pf fmt "[stmt] Sskip"
+  | Sset (id, e) ->
+      Fmt.pf fmt "[stmt] Sset (%d, %a)" (num_to_integer id) pp_expr e
+  | Sstore (chunk, e1, e2) ->
+      Fmt.pf fmt "[stmt] Sstore (%a, %a, %a)" Chunk.pp
+        (Chunk.of_compcert_ast_chunk chunk)
+        pp_expr e1 pp_expr e2
+  | Scall _ -> Fmt.pf fmt "[stmt] Scall"
+  | Sbuiltin _ -> Fmt.pf fmt "[stmt] Sbuiltin"
+  | Sseq (s1, s2) -> Fmt.pf fmt "[stmt] Sseq (%a, %a)" pp_stmt s1 pp_stmt s2
+  | Sifthenelse _ -> Fmt.pf fmt "[stmt] Sifthenelse "
+  | Sloop s -> Fmt.pf fmt "[stmt] Sloop (%a)" pp_stmt s
+  | Sblock s -> Fmt.pf fmt "[stmt] Sblock (%a)" pp_stmt s
+  | Sexit _ -> Fmt.pf fmt "[stmt] Sexit "
+  | Sswitch _ -> Fmt.pf fmt "[stmt] Sswitch "
+  | Sreturn _ -> Fmt.pf fmt "[stmt] Sreturn "
+  | Slabel (label, s) ->
+      Fmt.pf fmt "[stmt] Slabel (%d, %a)" (num_to_integer label) pp_stmt s
+  | Sgoto label -> Fmt.pf fmt "[stmt] Sgoto (%d)" (num_to_integer label)
+
+let rec trans_expr ~clight_prog ~fname ~fid ~local_env expr =
+  let trans_expr = trans_expr ~clight_prog ~fname ~fid ~local_env in
   let trans_binop_expr = trans_binop_expr ~fname in
   let gen_str = Generators.gen_str ~fname in
   let open Expr in
   match expr with
   | Evar id -> ([], PVar (true_name id))
   | Econst const -> ([], Lit (trans_const const))
-  | Eload (chunk, expp) ->
+  | Eload (compcert_chunk, expp) ->
       let cl, e = trans_expr expp in
       let gvar = gen_str Prefix.gvar in
       let loadv = Expr.Lit (Literal.String Internal_Functions.loadv) in
+
+      (* Lookup type of variable using Clight program to determine whether it is a pointer *)
+      let chunk =
+        match compcert_chunk with
+        (* Check whether chunk is a pointer - Mint64 or Mint32 can be pointer types *)
+        | Mint64 | Mint32 -> (
+            let typ =
+              match expp with
+              (* Get the id of the variable from CSharpMinor *)
+              | Ebinop (_, Evar id, _) ->
+                  let open Clight in
+                  let rec find_var var_list =
+                    match var_list with
+                    | (var_id, var_typ) :: vs ->
+                        if var_id == id then var_typ else find_var vs
+                    | _ ->
+                        failwith
+                          (Printf.sprintf
+                             "Variable with ident %d was not found inside \
+                              function %s of Clight"
+                             (num_to_integer id) fname)
+                  in
```

Comment:
`==` tests for physical equality in OCaml. But two equal ids are not guaranteed to be physically equal (as in one could be a copy of the other). Use `Camlcoq.P.eq` instead
Also,
```ocaml
let all_vars = clight_fun.fn_params @ clight_fun.fn_vars @ clight_fun.fn_temps in
let typ =
  match List.find_map (fun (var_id, var_typ) -> if Camlcoq.P.eq id var_id then Some var_typ else None) all_vars with
  | Some t -> t
  | None -> Fmt.failwith "Variable with ident %d ...." ...
in 
...
```

Reply:
That's very elegant